File: firestorm_flutter/lib/fs/queries/fs_query_result.dart
import 'package:cloud_firestore/cloud_firestore.dart';

/// Represents the result of a Firestore query, containing a list of objects and the snapshot of the last document, if any.
class FSQueryResult<T> {

  final List<T> _items;
  final QueryDocumentSnapshot? _snapshot;
  final String? _lastDocumentId;

  FSQueryResult(this._items, this._snapshot, this._lastDocumentId);

  String? get lastDocumentId => _lastDocumentId;

  List<T> get items => _items;

  QueryDocumentSnapshot? get snapshot => _snapshot;

}

File: firestorm-java/src/main/java/com/raylabz/firestorm/firestore/FSQueryResult.java
package com.raylabz.firestorm.firestore;

import com.google.cloud.firestore.QueryDocumentSnapshot;
import java.util.ArrayList;

/**
 * Models a result returned from a query.
 * @param <T> The type of objects in the result.
 * @author Nicos Kasenides
 * @version 1.0.0
 */
public class FSQueryResult<T> {

    private final ArrayList<T> items;
    private final QueryDocumentSnapshot snapshot;
    private final String lastDocumentID;

    /**
     * Constructs a query result.
     * @param items
     * @param snapshot
     * @param lastDocumentID
     */
    public FSQueryResult(ArrayList<T> items, QueryDocumentSnapshot snapshot, String lastDocumentID) {
        this.items = items;
        this.snapshot = snapshot;
        this.lastDocumentID = lastDocumentID;
    }

    public ArrayList<T> getItems() {
        return items;
    }

    public QueryDocumentSnapshot getSnapshot() {
        return snapshot;
    }

    public String getLastDocumentID() {
        return lastDocumentID;
    }

    public boolean hasItems() {
        return (!items.isEmpty());
    }

}
